### ISSUE TITLE
Redundant code after optimization (Clang10 vs Clang16)

### ISSUE THREAD

**Issue by ZY546** [llvm/llvm-project#64012]
Test code: https://godbolt.org/z/64aG9xz6e
```C++
int var_19,var_30;
void test(int var_0, int var_1, int var_2, int var_3, int var_4) {
    while(var_4){
        if(var_0){
            var_30 += var_1 + var_2;
        }
        else{
            var_19+=200/(var_1+var_2);
        }
        var_4++;
    }
}
```

IR generated by Clang10 -O1:
```
...
while.body.lr.ph:                                 ; preds = %entry
  %add = add nsw i32 %var_2, %var_1
  br label %while.body
...
```

However, IR generated by Clang16 -O1 with obvious redundancy of `%add3`:
```
...
while.body.lr.ph:                                 ; preds = %entry
  %var_19.promoted = load i32, ptr @var_19, align 4
  %var_30.promoted = load i32, ptr @var_30, align 4
  %add = add nsw i32 %var_2, %var_1
  %add3 = add nsw i32 %var_2, %var_1
  br label %while.body
...
```

**Comment by XChy** [comment]
I have no idea about why clang10 doesn't fold it. But in my knowledge, O1 in clang16 doesn't include `gvn`. Only with O2 can you eliminate such redundant code.

**Comment by ZY546** [comment]
> I have no idea about why clang10 folds it. But in my knowledge, O1 in clang16 doesn't include `gvn`. Only with O2 can you eliminate such redundant code.

But it might seem like an optimization regression, so I posed the question.

**Comment by XChy** [comment]
[It](https://godbolt.org/z/bjjnWvojx) shows that clang10 enables `HoistThenElseCodeToIf` in O1 pipeline by default, which folds this. Actually it has been disabled/moved to the end of pipeline by default in later LLVM version for many reasons.

**Comment by ZY546** [comment]
Is HoistThenElseCodeToIf an option for SimplifytheCFG? There seems to be no documentation on how I can enable or disable it. If yes, I'll look for Simplify the CFG's related files.

**Comment by XChy** [comment]
Yes. In new LLVM, HoistThenElseCodeToIf is an option disabled by default in SimplifyCFG Pass. The main reason is about vectorizations.
